contributors/design-proposals/storage/grow-flexvolume-size.md

```diff
@@ -0,0 +1,170 @@
+# Proposal for Growing FlexVolume Size
+
+**Authors:** [xingzhou](https://github.com/xingzhou)
+
+## Goals
+
+Since PVC resizing is introduced in Kubernetes v1.8, several volume plugins have already supported this feature, e.g. GlusterFS, AWS EBS. In this proposal, we are proposing to support FlexVolume expansion. So when user uses FlexVolume and corresponding volume driver to connect to his/her backend storage system, he/she can expand the PV size by updating PVC in Kubernetes.
+
+## Non Goals
+
+* We only consider expanding FlexVolume size in this proposal. Decreasing size of FlexVolume will be designed in the future.
+* In this proposal, user can only expand the FlexVolume size manually by updating PVC. Auto-expansion of FlexVolume based on specific meterings is not considered.
+* The proposal only contains the changes made in FlexVolume, volume driver changes which should be made by user are not included.
+
+## Implementation Designs
+
+### Prerequisites
+
+* Kubernetes should be at least v1.8.
+* Enable resizing by setting feature gate `ExpandPersistentVolumeGate` to `true`.
+* Enable `PersistentVolumeClaimResize` admission plugin(optional).
+* Follow the UI of PV resizing, including:
+  * Only dynamic provisioning supports volume resizing
+  * Set StorageClass attribute `allowVolumeExpansion` to `true`
+
+### Admission Control Changes
+
+Whether or not a specific volume plugin supports volume expansion is validated and checked in PV resize admission plugin. In general, we can list FlexVolume as the ones that support volume expansion and leave the actual expansion capability check to the underneath volume driver when PV resize controller calls the `ExpandVolumeDevice` method of FlexVolume.
```

Comment:
It looks like currently the resize admission controller has a hard-coded list of plugins that support resize. Is it worth augmenting it to dynamically check resize support?

Reply:
It is possible to do that, but there was a concern raised before that doing to will force us to load all volume plugins in admission controller. It may not be an issue for api-server since it *probably* already loads all volume plugins but this needs to be carefully considered. Last time I tried something like this - it ended up increasing  size of kubectl and few other binaries quite a bit.

Reply:
@saad-ali CSI will also likely run into this admission controller issue, right?

Reply:
I am thinking of removing static checking we are currently doing and not do any checks at all. The reasoning is - if admission controller is enabled, we only allow resizing of PVC created from storageClasses where `allowVolumeExpansion` property is set to true. Now if somehow - k8s admin enabled that field for a SC that does not support resizing, we will log an PVC event and that will enable k8s admin to fix the SC.

Or to rephrase - I think the fact that, we allow resizing of only dynamically provisioned PVC with `allowExpansion` set to `true` should be more than enough. if a k8s admin misconfigures the SC then it will be considered an admin error (and he should know better).